dense-numericals-src/basic-math/arg-maximum.lisp

```diff
@@ -129,26 +122,35 @@
                                           ((keep-dims null)))
     (signed-byte 64)
   (declare (ignore axis out keep-dims))
-  (pflet ((acc     -1)
+  (pflet ((val (type-min <type>))
           (c-name  (c-name <type> 'nu:arg-maximum))
+          (c-val-name  (c-name <type> 'nu:maximum))
           (c-size  (c-size <type>)))
-    (declare (type (signed-byte 64) acc))
-    (ptr-iterate-but-inner (narray-dimensions x)
-        n
-      ((ptr-x c-size inc-x x))
-      (setq acc
-            (cl:max acc (inline-or-funcall c-name n ptr-x inc-x))))
-    acc))
+    (declare (type <type> val))
+    (cl:let ((acc     -1)
+             (offset   0)
+             (stride (reduce #'* (cdr (narray-dimensions x)))))
+      (declare (cl:type (signed-byte 64) acc offset stride))
+      (ptr-iterate-but-inner (narray-dimensions x)
+          n
+        ((ptr-x c-size inc-x x))
+        (pflet ((new-val (inline-or-funcall c-val-name n ptr-x inc-x)))
+          (declare (type <type> new-val))
+          (when (> new-val val)
+            (setq acc (+ offset (inline-or-funcall c-name n ptr-x inc-x))
+                  val new-val)))
+        (incf offset stride))
```

Comment:
```lisp
(pflet ...)
```

This is a neat simple idea. Thanks for that! 

Not sure if I'm missing something, but perhaps we can eliminate the call to `c-val-name`: 

1. Compute the index of the local maximum/minimum using `c-name`.
2. Use it to directly access the local maximum/minimum without relying on `c-val-name`.

Reply:
Yes, I think that's better! Some problems I encounter while trying to implement this:
1. fixnum tests fail because cffi don't recognize `:fixnum`
2. I wrote some custom operators before using this approach and it seems `cffi:mem-ref` cannot optimize if `c-type` is bound to a compile time constant (?) by `pflet`. I had to resort to dirty hacks (a macro that `subst` in the body) to avoid the slow path for `cffi:mem-ref`. Is there a better way?

Reply:
Did you try [numericals/common:fref](https://github.com/digikar99/numericals/blob/1c236fc5606c1b8325ffb704f008b6ae3245ea7c/common.lisp#L197) instead of `cffi:mem-ref`? It might solve both issues. 

If not, we could connect over discord or telegram or let me know if there's another real-time communication channel you prefer. Otherwise github and github gists are fine too.

Reply:
> Did you try [numericals/common:fref](https://github.com/digikar99/numericals/blob/1c236fc5606c1b8325ffb704f008b6ae3245ea7c/common.lisp#L197) instead of `cffi:mem-ref`? It might solve both issues.

How to do that according to value of `<type>`? The issue of `cffi:mem-ref` falling to slow path happens for any types in general, not just to `fixnum`, which is due to its compiler macro not recognizing `<type>` as constant in e.g.  `(cffi:mem-ref array <type>)`.

> If not, we could connect over discord or telegram or let me know if there's another real-time communication channel you prefer. Otherwise github and github gists are fine too.

I'm happy to discuss! My telegram ID is the same as GitHub.

Reply:
Since `fref` is a polymorphic-function, its compiler macro can use information of the type of `<type>` from the (compiler)macro-expansion environment that `pflet` and `pf-compiler-macro` fill in. If that fails for some reason, I'd be happy to investigate; if my memory serves right, `cffi:mem-ref` failing to optimize beyond the trivial cases was one of the main reasons for writing `fref`. 

Also, just sent you a telegram ping. Same ID as Github on my end too.